### ISSUE TITLE
Assertion error in LDDWRdPtrQ pseudo expansion

### ISSUE THREAD

**Issue by dylanmckay** [avr-llvm/llvm#143]
The assertion that is triggered is located [here](https://github.com/avr-llvm/llvm/blob/7bdaae542e749df493f8ca35a9858d4b3805f01e/lib/Target/AVR/AVRExpandPseudoInsts.cpp#L794).

Error:
```
Assertion `DstReg != SrcReg && "SrcReg and DstReg cannot be the same"
```

The error can be reproduced with the `mul16.ll` CodeGen test.

[`mul16.ll`](https://github.com/avr-llvm/llvm/blob/7bdaae542e749df493f8ca35a9858d4b3805f01e/test/CodeGen/AVR/mul16.ll).

Command line:
``` bash
llc -march=avr -mcpu=atmega328p -filetype=obj -O0 -o a.out mul16.ll
```
Note that optimizations must be disabled, otherwise a different error occurs (as of 7bdaae542e749df493f8ca35a9858d4b3805f01e).

Program output:
```
llc: /home/dylan/projects/avr-llvm/llvm/lib/Target/AVR/AVRExpandPseudoInsts.cpp:794: bool llvm::AVRExpandPseudo::expand(llvm::AVRExpandPseudo::Block&, llvm::AVRExpandPseudo::BlockIt) [with unsigned int OP = 87u; llvm::AVRExpandPseudo::Block = llvm::MachineBasicBlock; llvm::AVRExpandPseudo::BlockIt = llvm::MachineBasicBlock::bundle_iterator<llvm::MachineInstr, llvm::ilist_iterator<llvm::MachineInstr> >]: Assertion `DstReg != SrcReg && "SrcReg and DstReg cannot be the same"' failed.
#0 0x132b127 llvm::sys::PrintStackTrace(llvm::raw_ostream&) /home/dylan/projects/avr-llvm/llvm/lib/Support/Unix/Signals.inc:437:0
#1 0x132b4a4 PrintStackTraceSignalHandler(void*) /home/dylan/projects/avr-llvm/llvm/lib/Support/Unix/Signals.inc:495:0
#2 0x132a07b SignalHandler(int) /home/dylan/projects/avr-llvm/llvm/lib/Support/Unix/Signals.inc:210:0
#3 0x7f13d276e660 __restore_rt (/usr/lib/libpthread.so.0+0x10660)
#4 0x7f13d193d528 __GI_raise (/usr/lib/libc.so.6+0x33528)
#5 0x7f13d193e93a __GI_abort (/usr/lib/libc.so.6+0x3493a)
#6 0x7f13d19363a7 __assert_fail_base (/usr/lib/libc.so.6+0x2c3a7)
#7 0x7f13d1936452 (/usr/lib/libc.so.6+0x2c452)
#8 0x9ece5d bool llvm::AVRExpandPseudo::expand<87u>(llvm::MachineBasicBlock&, llvm::MachineBasicBlock::bundle_iterator<llvm::MachineInstr, llvm::ilist_iterator<llvm::MachineInstr> >) /home/dylan/projects/avr-llvm/llvm/lib/Target/AVR/AVRExpandPseudoInsts.cpp:799:0
#9 0x9efde5 llvm::AVRExpandPseudo::expandMI(llvm::MachineBasicBlock&, llvm::MachineBasicBlock::bundle_iterator<llvm::MachineInstr, llvm::ilist_iterator<llvm::MachineInstr> >) /home/dylan/projects/avr-llvm/llvm/lib/Target/AVR/AVRExpandPseudoInsts.cpp:1391:0
#10 0x9e9227 llvm::AVRExpandPseudo::expandMBB(llvm::MachineBasicBlock&) /home/dylan/projects/avr-llvm/llvm/lib/Target/AVR/AVRExpandPseudoInsts.cpp:81:0
#11 0x9e9307 llvm::AVRExpandPseudo::runOnMachineFunction(llvm::MachineFunction&) /home/dylan/projects/avr-llvm/llvm/lib/Target/AVR/AVRExpandPseudoInsts.cpp:97:0
#12 0xbd3eed llvm::MachineFunctionPass::runOnFunction(llvm::Function&) /home/dylan/projects/avr-llvm/llvm/lib/CodeGen/MachineFunctionPass.cpp:41:0
#13 0xf89a7f llvm::FPPassManager::runOnFunction(llvm::Function&) /home/dylan/projects/avr-llvm/llvm/lib/IR/LegacyPassManager.cpp:1520:0
#14 0xf89bde llvm::FPPassManager::runOnModule(llvm::Module&) /home/dylan/projects/avr-llvm/llvm/lib/IR/LegacyPassManager.cpp:1540:0
#15 0xf89f0d (anonymous namespace)::MPPassManager::runOnModule(llvm::Module&) /home/dylan/projects/avr-llvm/llvm/lib/IR/LegacyPassManager.cpp:1596:0
#16 0xf8a56e llvm::legacy::PassManagerImpl::run(llvm::Module&) /home/dylan/projects/avr-llvm/llvm/lib/IR/LegacyPassManager.cpp:1698:0
#17 0xf8a773 llvm::legacy::PassManager::run(llvm::Module&) /home/dylan/projects/avr-llvm/llvm/lib/IR/LegacyPassManager.cpp:1730:0
#18 0x9c9aa5 compileModule(char**, llvm::LLVMContext&) /home/dylan/projects/avr-llvm/llvm/tools/llc/llc.cpp:367:0
#19 0x9c8c1d main /home/dylan/projects/avr-llvm/llvm/tools/llc/llc.cpp:204:0
#20 0x7f13d192a790 __libc_start_main (/usr/lib/libc.so.6+0x20790)
#21 0x9c7a29 _start (/home/dylan/projects/builds/avr-llvm/llvm/bin/llc+0x9c7a29)
Stack dump:
0.      Program arguments: ./llc -march=avr -mcpu=atmega328p -filetype=obj -O0 -o a.out /home/dylan/projects/avr-llvm/llvm/test/CodeGen/AVR/mul16.ll 
1.      Running pass 'Function Pass Manager' on module '/home/dylan/projects/avr-llvm/llvm/test/CodeGen/AVR/mul16.ll'.
2.      Running pass 'AVR pseudo instruction expansion pass' on function '@_Z3maplllll'
[1]    8116 abort (core dumped)  ./llc -march=avr -mcpu=atmega328p -filetype=obj -O0 -o a.out 

```

**Comment by dylanmckay** [comment]
Closing, duplicate of #140.
